examples/latent_sde.py

```diff
@@ -71,23 +71,46 @@ def f(self, t, y):  # Approximate posterior drift.
         inp = torch.cat((torch.sin(t), torch.cos(t), y), dim=-1)
         return self.net(inp)
 
+    def f_aug(self, t, y):
+        y = y[:, 0:1]
+        f, g, h = self.f(t, y), self.g(t, y), self.h(t, y)
+        z = f - h
+        g = torch.where(
+            g.abs() > 1e-7,
+            g,
+            torch.ones_like(g).fill_(1e-7) * g.sign()
+        )
+        z = z / g
+        z = .5 * (torch.norm(z, dim=1, keepdim=True) ** 2.)
+        return torch.cat([f, z], dim=1)
+
     def g(self, t, y):  # Shared diffusion.
         return self.sigma.repeat(y.size(0), 1)
 
+    def g_aug(self, t, y):
+        y = y[:, 0:1]
+        g = self.g(t, y)
+        z = torch.zeros_like(y)
+        return torch.cat([g, z], dim=1)
+
     def forward(self, ts, batch_size, eps=None):
-        eps = torch.randn(batch_size, 1).to(self.qy0_std) if eps is None else eps
+        eps = torch.cat([torch.randn(batch_size, 1).to(self.qy0_std), torch.zeros(batch_size, 1)
+                        .to(self.qy0_std)], dim=1) if eps is None else eps
```

Comment:
This part is a little weird, since the second entry of y0 is likely not 0 (it seems to be `qy0_mean`), due to tensor broadcasting. 

Additionally, the KL divergence compute is also weird. 

This should be able to be modified by computing `eps`, `y0`, and the KL as before, but "append" y0 with zeros on the extra logqp dimension just before starting `sdeint`. 

I think this might have been a major cause in the bad performance.

Reply:
OK, so regarding `y0` it was me completely not paying attention to the code 🙄 - fixed it. 

Right now result is almost identical to `master` version (after 200 iterations):

![](https://iili.io/22Yqhu.png)

Also after above explanation of computing subintervals and examining it once again I think now I fully understand `latent_sde.py` - so subintervals define which points we want after all from the whole solve and those here are the `x`s from `cos` training data. Then in loss we check how much prediction is off from `cos` data and we don't want to do it with just MSE because according to the paper it would cause overfitting (is it correct?), instead we check probability density function with mean at predicted values at `x`s `cos` data. 

But only fitting to the data it would make it arbitrary between data points (correct?). And here's such case after 500 iters:

![](https://iili.io/22mBDJ.png)

Instead we want it to behave as Ornstein–Uhlenbeck process while fitting the data, so we compute through the whole solve how much our solution still behave like OU process (and regularize it that way). 

To compute the difference we use KL divergence to see how much they differ.
We start computing KL div between mean and std params of prior and posterior and then within solve - accumulate KL div for each step where KL div `dQ/dP` from Girsanov theorem will be Doléans-Dade exponential where it's `u-1/2*|u|^2`. But I can't see where `u = (f-h)/g` came from and how it would look like if prior and posterior had different diffusions, and why `u` is missing from that KL div step definiton. Do you have a link to the theory where I can look for it?

That's added to the loss, and we also decrease learning rate to make training process faster at the beginning. 

Could be `py0_mean` and `py0_logvar` be just tensors as we do not train them?

Reply:
There are easier and more intuitive ways to derive this, but for a formal derivation, you can checkout section 9.5 of our paper.